File: Python/textbook_track/chapter07/ba7a.py
# Compute Distances Between Leaves
import sys

from collections import defaultdict
from math import inf


def read_weighted_edges(lines):
    adjacency_list = defaultdict(list)
    for line in lines:
        node_from, node_to, weight = map(int, line.replace('->', ':').split(':'))
        adjacency_list[node_from].append((node_to, weight))
    return dict(adjacency_list)


def generate_nodes(adjacency_list):
    for node, neighbours in adjacency_list.items():
        for neighbour, _ in neighbours:
            yield neighbour


def initialize_distance_matrix(adjacency_list, nr_nodes):
    distances = [[inf] * nr_nodes for _ in range(nr_nodes)]
    for node, neighbours in adjacency_list.items():
        for neighbour, weight in neighbours:
            distances[node][neighbour] = weight

    for node in range(nr_nodes):
        distances[node][node] = 0

    return distances


def calc_pairwise_distances(adjacency_list):
    nodes = generate_nodes(adjacency_list)
    nr_nodes = max(nodes) + 1
    distances = initialize_distance_matrix(adjacency_list, nr_nodes)
    for k in range(nr_nodes):
        for ix in range(nr_nodes):
            for jy in range(nr_nodes):
                distances[ix][jy] = min(distances[ix][jy], distances[ix][k] + distances[k][jy])
    return distances


def compute_distances_between_leaves(adjacency_list, nr_leaves):
    distances = calc_pairwise_distances(adjacency_list)
    return [row[:nr_leaves] for row in distances[:nr_leaves]]


def main():
    data = sys.stdin.read().splitlines()
    nr_leaves = int(data[0])
    adjacency_list = read_weighted_edges(data[1:])
    result = compute_distances_between_leaves(adjacency_list, nr_leaves)
    for row in result:
        print(' '.join(map(str, row)))


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter07/BA7A.scala
package TextbookTrack.Chapter07

object BA7A {
  final case class WeightedEdge(nodeFrom: Int, nodeTo: Int, weight: Int)

  class Tree(val adjacencyList: Map[Int, List[WeightedEdge]], val nrLeaves: Int) {
    lazy val nrNodes: Int = adjacencyList.valuesIterator.flatMap(_.map(_.nodeTo)).max + 1

    def calcPairwiseDistances(): List[List[Int]] = {
      val distances: Array[Array[Double]] = initializeDistanceMatrix()
      for {
        k <- 0 until nrNodes
        ix <- 0 until nrNodes
        jy <- 0 until nrNodes
      } distances(ix)(jy) = math.min(distances(ix)(jy), distances(ix)(k) + distances(k)(jy))

      distances.map(row => row.map(_.toInt).toList).toList
    }

    private def initializeDistanceMatrix(): Array[Array[Double]] = {
      val distances: Array[Array[Double]] = Array.fill(nrNodes, nrNodes)(Double.PositiveInfinity)
      for {
        (node, neighbours) <- adjacencyList
        WeightedEdge(_, neighbour, weight) <- neighbours
      } distances(node)(neighbour) = weight

      (0 until nrNodes).foreach(node => distances(node)(node) = 0)

      distances
    }
  }

  private def splitEdge(line: String): WeightedEdge = line.replace("->", ":").split(":").map(_.toInt).toList match {
    case List(from, to, weight) => WeightedEdge(from, to, weight)
    case _ => throw new Exception("Malformed input data.")
  }

  def readWeightedEdges(lines: Iterator[String]): Map[Int, List[WeightedEdge]] =
    lines.map(splitEdge).toList.groupBy(_.nodeFrom)

  def computeDistancesBetweenLeaves(tree: Tree): List[List[Int]] = {
    val distances: List[List[Int]] = tree.calcPairwiseDistances()
    distances.take(tree.nrLeaves).map(_.take(tree.nrLeaves))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrLeaves: Int = reader.next().toInt
    val adjacencyList: Map[Int, List[WeightedEdge]] = readWeightedEdges(reader)
    val tree = new Tree(adjacencyList, nrLeaves)
    val result: List[List[Int]] = computeDistancesBetweenLeaves(tree)
    result.foreach(line => println(line.mkString(" ")))
  }
}
